src/ctx.rs

```diff
@@ -1,147 +1,240 @@
-/// or else skipped if `WITH_DEFAULT` is `false`.
+/// A more realistic example of filling multiple buffers with the same value is
+/// initializing different struct fields at the same time (from
+/// `src/decode.rs`):
+/// ```ignore
+/// case_set!(32, ctx=[(&t.l, 1), (&f.a[t.a], 0)], len=[bh4, bw4], offset=[by4, bx4], {
+///      let (dir, dir_index) = ctx;
+///      set_disjoint!(dir.seg_pred, seg_pred.into());
+///      set_disjoint!(dir.skip_mode, b.skip_mode);
+///      set_disjoint!(dir.intra, 0);
+///      set_disjoint!(dir.skip, b.skip);
+///      set_disjoint!(dir.pal_sz, 0);
+/// });
+/// ```
 ///
-/// This optimizes for the common cases where `buf.len()` is a small power of 2,
-/// where the array write is optimized as few and large stores as possible.
-#[inline]
-pub fn small_memset<T: Clone + Copy, const UP_TO: usize, const WITH_DEFAULT: bool>(
-    buf: &mut [T],
-    val: T,
-) {
-    fn as_array<T: Clone + Copy, const N: usize>(buf: &mut [T]) -> &mut [T; N] {
-        buf.try_into().unwrap()
-    }
-    match buf.len() {
-        01 if UP_TO >= 01 => *as_array(buf) = [val; 01],
-        02 if UP_TO >= 02 => *as_array(buf) = [val; 02],
-        04 if UP_TO >= 04 => *as_array(buf) = [val; 04],
-        08 if UP_TO >= 08 => *as_array(buf) = [val; 08],
-        16 if UP_TO >= 16 => *as_array(buf) = [val; 16],
-        32 if UP_TO >= 32 => *as_array(buf) = [val; 32],
-        64 if UP_TO >= 64 => *as_array(buf) = [val; 64],
-        _ => {
-            if WITH_DEFAULT {
-                buf.fill(val)
+/// [`DisjointMut`]: crate::src::disjoint_mut::DisjointMut
+macro_rules! case_set {
+    ($UP_TO:literal, $(@DEFAULT=$WITH_DEFAULT:literal,)? $ctx:ident=[$($ctx_expr:expr),* $(,)?], $len:ident=[$($len_expr:expr),* $(,)?], $offset:ident=[$($offset_expr:expr),* $(,)?], $body:block) => {
+        let ctxs = [$($ctx_expr,)*];
+        let lens = [$($len_expr,)*];
+        let offsets = [$($offset_expr,)*];
+        assert_eq!(ctxs.len(), lens.len());
+        assert_eq!(ctxs.len(), offsets.len());
+        for (i, ctx) in ctxs.into_iter().enumerate() {
+            case_set!($UP_TO, $(@DEFAULT=$WITH_DEFAULT,)? $ctx=ctx, $len=lens[i], $offset=offsets[i], $body);
+        }
+    };
+    ($UP_TO:literal, $(@DEFAULT=$WITH_DEFAULT:literal,)? $len:ident, $offset:ident, $body:block) => {
+        case_set!($UP_TO, $(@DEFAULT=$WITH_DEFAULT,)? _ctx=(), $len=$len, $offset=$offset, $body);
+    };
+    ($UP_TO:literal, $(@DEFAULT=$WITH_DEFAULT:literal,)? $len:ident=$len_expr:expr, $offset:ident=$offset_expr:expr, $body:block) => {
+        case_set!($UP_TO, $(@DEFAULT=$WITH_DEFAULT,)? _ctx=(), $len=$len_expr, $offset=$offset_expr, $body);
+    };
+    ($UP_TO:literal, $(@DEFAULT=$WITH_DEFAULT:literal,)? $ctx:ident=$ctx_expr:expr, $len:ident=$len_expr:expr, $offset:ident=$offset_expr:expr, $body:block) => {
+        #[allow(unused_mut)]
+        let mut $ctx = $ctx_expr;
+        let $len = $len_expr;
+        let $offset = $offset_expr;
```

Comment:
Do these need to be variable names set by the caller?  I don't think they're accessed in any `$body:block`.

Reply:
Then we don't have to worry about name collisions like `offset` and `case_offset`.

Reply:
Only `$ctx` is accessed by the caller IIUC.

Reply:
I see this now, good point. I figured somewhere we access it somewhere from looking at the original C macro, but I guess we don't. It's a bit confusing that the body can access the context variable but not the others. I think I'd prefer to keep it consistent even if we don't use it right now.

I guess we could use `let ctx = ...` instead to indicate that it creates a new variable name, if we really want to make them different.

Reply:
In the old API, though, `ctx` was explicitly the only extra variable passed to the closure.  I don't think it's confusing that the others aren't accessible.  `ctx` is special in this way.  It's more confusing to me that there's all these other variables the caller defines.